chain/src/Pos/Chain/Block/Logic/Integrity.hs

```diff
@@ -166,13 +170,27 @@ verifyHeader pm VerifyHeaderParams {..} h =
         [ checkDifficulty
               (prevHeader ^. difficultyL + headerDifficultyIncrement h)
               (h ^. difficultyL)
+        -- This \/ checks that the previous block's header matches what our current block
+        -- says the previous header hash was. This verifies integrity of the chain.
         , checkHash
               (headerHash prevHeader)
               (h ^. prevBlockL)
         , checkEpochOrSlot (getEpochOrSlot prevHeader) (getEpochOrSlot h)
         , case h of
               BlockHeaderGenesis _ -> (True, "") -- check that epochId prevHeader < epochId h performed above
-              BlockHeaderMain _    -> sameEpoch (prevHeader ^. epochIndexL) (h ^. epochIndexL)
+              BlockHeaderMain _    -> case vhpConsensusEra of
+                Original -> sameEpoch (prevHeader ^. epochIndexL) (h ^. epochIndexL)
+                OBFT _   -> case unEpochOrSlot (getEpochOrSlot h) of
+                    Left _ -> (True, "we received an Epoch Boundary Block in OBFT era, which shouldn't \
+                                     \happen, but we are passing")
+                    Right sid -> case (getEpochIndex (siEpoch sid), getSlotIndex (siSlot sid)) of
+                        (0, 0) -> (True, "first MainBlock of first epoch: no previous block")
+                        -- First block of the epoch: assert its epoch is 1 greater than
+                        -- the epoch of the prior block.
+                        (_, 0) -> sameEpoch (1 + prevHeader ^. epochIndexL) (h ^. epochIndexL)
```

Comment:
Note to self: 

Looks like this check could be problematic as the first block of the epoch might not occur at the 0th slot. For example, if the 0th slot leader is offline and unable to create a block, the first block of that epoch will not occur at slot 0 since it will be skipped.